addons/website/models/ir_asset.py

```diff
@@ -10,35 +10,48 @@ class IrAsset(models.Model):
     key = fields.Char(copy=False) # used to resolve multiple assets in a multi-website environment
     website_id = fields.Many2one('website', ondelete='cascade')
 
-    def _get_related_assets(self, domain):
-        website = self.env['website'].get_current_website(fallback=False)
-        if website:
-            domain += website.website_domain()
-        assets = super()._get_related_assets(domain)
-        return assets.filter_duplicate()
-
-    def _get_active_addons_list(self):
+    def _get_asset_params(self):
+        params = super()._get_asset_params()
+        params['website_id'] = self.env['website'].get_current_website(fallback=False).id
+        return params
+
+    def _get_asset_extra(self, extra, website_id=None, **params):
+        extra = super()._get_asset_extra(extra, **params)
+        if extra == '%':
+            return extra
+        website_id_path = website_id and ('%s/' % website_id) or ''
+        return website_id_path + extra
+
+    def _get_related_assets(self, domain, website_id=None, **params):
+        if website_id:
+            domain += self.env['website'].website_domain(website_id)
+        assets = super()._get_related_assets(domain, **params)
+        return assets.filter_duplicate(website_id)
```

Comment:
As the base implementation is the following:
```py
    def _get_related_assets(self, domain):
        """
        Returns a set of assets matching the domain, regardless of their
        active state. This method can be overridden to filter the results.
        :param domain: search domain
        :returns: ir.asset recordset
        """
        # active_test is needed to disable some assets through filter_duplicate for website
        # they will be filtered on active afterward
        return self.with_context(active_test=False).sudo().search(domain, order='sequence, id')
```

Wouldn't it be better to add `**kw` on this base implementation and keep the same signature here, poping `website_id` from it before calling `super()`?

The different signature between those override seems weird to me, especially once other modules will / might start writing their own different one too.

Or is it actually a good practice / pattern the way it is currently implemented?

Reply:
It is subjective I think I was a little `partagé` between the two solutions but my linter indicates when I have a unused parameter but not when the ovverrides have different signatures. And I actually like this pattern. But juc will answer in details why.

Reply:
> The different signature between those override seems weird to me

It is normal you find it weird, this is a way to implement method *overload* (=adding a new signature to a method in a child class) in Python. There are two ways to do overload, (1) change the type of one of the arguments (e.g. list vs tuple), (2) add new arguments. (1) is much more common than (2) in the wild and it is solved generally solved in Python thanks to duck-typing (usually [datamodel](https://docs.python.org/3/reference/datamodel.html) + [collection.abc](https://docs.python.org/3/library/collections.abc.html)) meaning we usually don't need to do any Java-like type-1 overload. Here we wanted a type-2 overload, those are much less common, specially in Python that as no dedicated syntax for it unlike Java, what you are reading here is a way to implement type-2 overload.

> might start writing their own different one too

This is actually part of the pattern, we *enable* method overload on that method, other classes can overload it their own way and add their new parameters. This is expected.

> Or is it actually a good practice

This is not good/bad practice, this is one way of implementing method overload in Python. Are there other solutions than method overload to solve this problem: surely, would they be better: maybe, is the current solution good/simple enough: we think so

> Wouldn't it be better to add **kw on this base implementation and keep the same signature here, poping website_id from it before calling super()?

This would work too and we would had implemented it this way:

```py
def _get_related_assets(self, domain, **kw: dict):
    webside_id = kw.pop('webside_id', None)
    super()._get_related_assets(domain, **kw)
    ...

def _get_related_assets(self, domain, **kw: dict):
    if kw:
        raise TypeError(f"The base implementation takes no extra keyword arguments, but some are present. Thoses KW should had been consumed by overrides: {kw}")
    ...
```

The benefit of using the current implementation using method overload is that all the extra kwargs that will be used within the method are explicit on the function signature. It is also explicit that the base implementation takes no extra kwargs and the `raise TypeError(...)` is automatically done by Python.

Depending on whether an override/overload will/will not *consume* (use it and remove it from the kwargs),

Override:
```py
def foo(**kwargs):
    x = kwargs.get("x", None)  # does not consume
    y = kwargs.pop("y", None)  # does consume
    super().foo(**kwargs)
```

Overload:
```py
def foo(x, y, **kwargs):
    super().foo(x, **kwargs)
```

---

If you do not like this pattern and rather we implement a more classic override (like in the Override example) it is your call, we basically implemented it only for website :p

Reply:
> This would work too and we would had implemented it this way:

That's what I had in mind/suggested in my previous message indeed, except for the "raise no kw", is that really needed?

> If you do not like this pattern and rather we implement a more classic override (like in the Override example) it is your call, we basically implemented it only for website :p

No it's fine, it's not usual, at least in Odoo and bugged me when reading it, also the part where you need to remove params when calling the super, not that straightforward.
So I was more curious than anything else.
Thanks for the detailed explanation, don't need to change the implementation :+1:

Reply:
> you need to remove params when calling the super, not that straightforward

yes you are right :+1:, this is uncommon to have a super() call that is different from the function signature

Reply:
As juc said since it is mainly for website I have no issue to change that (but as you said without the raise). It is just a pattern I like but there is no other valid rational reason to use it except maybe for the current linting rules that are in place.

Reply:
I made a quick branch to show the difference with another pattern:

https://github.com/odoo-dev/odoo/compare/master-refactor-assets-ormcache-xdo..master-example-pattern-xdo?expand=1

One drawback of this solution is that you may need to modify all overrides if you want to add any param to the _get_asset_params method.

We can imagine that it will only have an effect on _get_related_assets meaning that the
_get_active_addons_list would be useless and you will need to do it. But actually, the _get_asset_extra would need to me modified because it is not correct to change the behavior of generation of assets based on a new parameter without adding this info in the extra so it looks like a good idea to force this override.

Anyway, I don't think anyone else than website will override this method. The whole _get_asset_params is maybe overkill since we only need the website_id. But base is not aware of that unfortunately. I think this implementation is a good compromise between extensibility and readability, even if how it works can look a little complicated, almost wrong from the outside when looking closely.

Reply:
> I made a quick branch to show the difference with another pattern:
>
> https://github.com/odoo-dev/odoo/compare/master-refactor-assets-ormcache-xdo..master-example-pattern-xdo?expand=1

Note that I would've gotten rid of the `website_id` in the signature and have it inside the `**kw` but as you said, don't worry about it, I got the explanation I am fine with it :)